File: ssd_app/my_custom/report/shipping_book_table/shipping_book_table.py
# Copyright (c) 2026, SSDolui and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import formatdate, flt
from frappe.utils.pdf import get_pdf
from frappe.utils.jinja import render_template
from frappe import _


def get_cif_data(filters):
    year = filters.year
    limit = filters.get("limit")
    status = filters.get("status")
    
    limit_clause = ""
    if limit and int(limit) > 0: 
        limit_clause = f"LIMIT {int(limit)}"

    conditional_filter = "WHERE 1=1"
    if not year:
        max_year = frappe.db.sql("""
            SELECT MAX(YEAR(bl_date))
            FROM `tabShipping Book`
            WHERE bl_date IS NOT NULL
        """, as_list=True)[0][0]
        conditional_filter += f""" AND YEAR(sb.bl_date)= {int(max_year)}"""
    else:
        conditional_filter += f""" AND YEAR(sb.bl_date)= {int(year)}"""
        
    if(status =="cif_pending"):
        conditional_filter += " AND cif.name IS NULL "
    elif(status =="cif_done"):
        conditional_filter += " AND cif.name IS NOT NULL "
        
    data= frappe.db.sql(f"""
        SELECT sb.name, sb.inv_no, sb.bl_date, com.company_code AS s_com, cus.customer, noti.notify, sb.invoice_amount, sb.document, 
                        term.term_name AS p_term, sb.term_days, bank.bank, user.first_name AS created_by, cif.name AS cif_id, "-" AS action
FROM `tabShipping Book` sb
LEFT JOIN `tabCompany` com ON com.name= sb.company
LEFT JOIN `tabCustomer` cus ON cus.name= sb.customer
LEFT JOIN `tabNotify` noti ON noti.name= sb.notify
LEFT JOIN `tabPayment Term` term ON term.name= sb.payment_term
LEFT JOIN `tabBank` bank ON bank.name= sb.bank
LEFT JOIN `tabUser` AS user ON sb.owner= user.name
LEFT JOIN `tabCIF Sheet` cif ON cif.inv_no= sb.name

    {conditional_filter}
    ORDER BY sb.creation DESC 
    {limit_clause};
    """, as_dict=1)
    return data


def execute(filters=None):
    columns = [
        # {"label": "Inv ID", "fieldname": "name", "fieldtype": "Data", "width": 80},
        {"label": "Inv No", "fieldname": "inv_no", "fieldtype": "Data", "width": 90},
        {"label": "BL Date", "fieldname": "bl_date", "fieldtype": "Date", "width": 110},
        {"label": "Shi Com", "fieldname": "s_com", "fieldtype": "Data", "width": 110},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Data", "width": 180},
        {"label": "Notify", "fieldname": "notify", "fieldtype": "Data", "width": 220},
        {"label": "Invoice", "fieldname": "invoice_amount", "fieldtype": "Float", "width": 100},
        {"label": "Document", "fieldname": "document", "fieldtype": "Float", "width": 100},
        {"label": "Bank", "fieldname": "bank", "fieldtype": "Data", "width": 60},
        {"label": "P Term", "fieldname": "p_term", "fieldtype": "Data", "width": 100},
        {"label": "T Days", "fieldname": "term_days", "fieldtype": "Data", "width": 80},
        {"label": "Create By", "fieldname": "created_by", "fieldtype": "Data", "width": 80},
        {"label": "Action", "fieldname": "action", "fieldtype": "Data", "width": 80}
    ]
    data = get_cif_data(filters)
    return columns, data

@frappe.whitelist()
def get_years():
    years = frappe.db.sql("""
        SELECT DISTINCT YEAR(bl_date) AS year
        FROM `tabShipping Book`
        WHERE bl_date IS NOT NULL
        ORDER BY year ASC
    """, as_dict=True)

    # Return a simple list of years as strings
    return [str(d.year) for d in years]

File: ssd_app/my_custom/report/shipping_book_table/shipping_book_table.js
// Copyright (c) 2026, SSDolui and contributors
// For license information, please see license.txt


if (typeof html2pdf === "undefined") {
    const script = document.createElement("script");
    script.src = "https://cdnjs.cloudflare.com/ajax/libs/html2pdf.js/0.10.1/html2pdf.bundle.min.js";
    script.defer = true;
    document.head.appendChild(script);
}


frappe.query_reports["Shipping Book Table"] = {
    onload: function (report) {
        report.page.add_inner_button("CIF Sheet Table", function () {
            frappe.set_route("query-report", "CIF Sheet Table");
        });
        // report.page.add_inner_button("Open CIF Sheet List", function () {
        //     frappe.set_route("List", "CIF Sheet");
        // });
        report.page.add_inner_button(__("Create New"), function () {
            // Set the flag in sessionStorage before navigating
            sessionStorage.setItem('return_to_after_save', 'Shipping Book Table');
            frappe.new_doc("Shipping Book");
        });

        report.refresh();

        frappe.call({
            method: "ssd_app.my_custom.report.shipping_book_table.shipping_book_table.get_years",
            callback: function (r) {
                if (r.message) {
                    let year_filter = report.get_filter("year");
                    let years = r.message;

                    // Set default as MAX year
                    year_filter.df.default = Math.max(...years.map(y => parseInt(y))).toString();

                    // Add "All" option at the beginning
                    years.unshift("All");

                    // Set options
                    year_filter.df.options = years;

                    // Refresh to apply default & options
                    year_filter.refresh();
                    year_filter.set_input(year_filter.df.default);
                }
            }
        });
    },
    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

                // Action Column
        if (column.fieldname === "action" && value) {

            let cif_button = "";

            if (!data.cif_id) {
                cif_button = `
                    <a href="#"
                        onclick="create_cif_sheet('${data.name}'); return false;"
                        title="Create CIF Sheet">
                        <svg class="icon icon-sm">
                            <use href="#icon-upload"></use>
                        </svg>
                    </a>
                `;
            }

            return `
                <div style="display:flex; gap:12px; align-items:center;">

                    <!-- Edit -->
                    <a href="/app/shipping-book/${data.name}"
                        onclick="sessionStorage.setItem('return_to_after_save', 'Shipping Book Table')"
                        title="Edit">
                        <svg class="icon icon-sm">
                            <use href="#icon-edit"></use>
                        </svg>
                    </a>

                    ${cif_button}

                </div>
            `;
        }


        // Clickable inv_no with modal
        if (column.fieldname === "inv_no" && data && data.cif_id) {
            return `<a href="#" onclick="showCIFDetails('${data.cif_id}', '${data.inv_no}'); return false;">${data.inv_no}</a>`;
        }

        return value;
    },
	filters: [   
        {
            fieldname: "year",
            label: "Year",
            fieldtype: "Select",
            options: [],   // will be filled dynamically
            reqd: 0
        },
        {
            fieldname: "limit",
            label: __("Limit"),
            fieldtype: "Select",
            options: [
                { "value": 100, "label": __("100") },
                { "value": 500, "label": __("500") },
                { "value": 0, "label": __("All") } // Use 0 or "" to represent 'No Limit' in your query
            ],
            default: 100,
            reqd: 0
        },
		{
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: [
                { "value": "all", "label": __("All") },
                { "value": "cif_pending", "label": __("CIF Pending") },
                { "value": "cif_done", "label": __("CIF Done") } // Use 0 or "" to represent 'No Limit' in your query
            ],
            default: "all",
            reqd: 0
        },
        {
            fieldname: "quick_search",
            label: __("Quick Search"),
            fieldtype: "Data",
            on_change: function() {
                const report = frappe.query_report;
                const search_term = (this.get_value() || "").toLowerCase();
                
                if (!report || !report.datatable) return;

                const datatable = report.datatable;
                const all_rows = datatable.datamanager.getRows();

                if (!search_term) {
                    // Reset to show everything
                    datatable.rowmanager.showRows(all_rows.map((_, i) => i));
                } else {
                    // Find indices of rows that match
                    const matches = all_rows
                        .map((row, index) => {
                            // Check every cell in the row
                            const has_match = row.some(cell => {
                                const val = String(cell.content || "").toLowerCase();
                                return val.includes(search_term);
                            });
                            return has_match ? index : null;
                        })
                        .filter(idx => idx !== null);

                    datatable.rowmanager.showRows(matches);
                }

                // Essential: Update the display dimensions and refresh
                datatable.dimensions.recompute();
                datatable.refresh();
            }
        }
    ],
};


function create_cif_sheet(name) {
    // frappe.return_to_page='CIF Sheet Table';
    sessionStorage.setItem(
        'return_to_page',
        'CIF Sheet Table'
    );

    frappe.new_doc("CIF Sheet");

    setTimeout(() => {
        cur_frm.set_value("inv_no", name);
    }, 200);
}

